File: Classes/TiGooglemapsPolylineProxy.m
/**
 * Ti.GoogleMaps
 * Copyright (c) 2015-present by Hans Knöchel. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsPolylineProxy.h"

@implementation TiGooglemapsPolylineProxy

- (GMSPolyline *)polyline
{
  if (_polyline == nil) {
    _polyline = [GMSPolyline new];
    _polyline.tappable = YES;
  }

  return _polyline;
}

- (void)setPoints:(id)points
{
  ENSURE_UI_THREAD(setPoints, points);
  ENSURE_TYPE_OR_NIL(points, NSArray);

  [self replaceValue:points forKey:@"points" notification:NO];

  GMSMutablePath *path = [GMSMutablePath path];

  if (points != nil) {
    if ([points count] < 2) {
      NSLog(@"[WARN] Ti.GoogleMaps: You need to specify at least 2 points to create a polygon.");
      return;
    }

    for (id point in points) {
      if ([point isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point valueForKey:@"latitude"]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point valueForKey:@"longitude"]];

        [path addLatitude:latitude longitude:longitude];
      } else if ([point isKindOfClass:[NSArray class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point objectAtIndex:1]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point objectAtIndex:0]];

        [path addLatitude:latitude longitude:longitude];
      }
    }
  }

  [[self polyline] setPath:path];
}

- (void)setTappable:(id)value
{
  ENSURE_UI_THREAD(setTappable, value);
  ENSURE_TYPE(value, NSNumber);

  [[self polyline] setTappable:[TiUtils boolValue:value]];
  [self replaceValue:value forKey:@"tappable" notification:NO];
}

- (void)setStrokeColor:(id)value
{
  ENSURE_UI_THREAD(setStrokeColor, value);

  [[self polyline] setStrokeColor:[[TiUtils colorValue:value] _color]];
  [self replaceValue:value forKey:@"strokeColor" notification:NO];
}

- (void)setStrokeWidth:(id)value
{
  ENSURE_UI_THREAD(setStrokeWidth, value);

  [[self polyline] setStrokeWidth:[TiUtils floatValue:value def:1]];
  [self replaceValue:value forKey:@"strokeWidth" notification:NO];
}

- (void)setGeodesic:(id)value
{
  ENSURE_UI_THREAD(setGeodesic, value);

  [[self polyline] setGeodesic:[TiUtils boolValue:value def:NO]];
  [self replaceValue:value forKey:@"geodesic" notification:NO];
}

- (void)setTitle:(id)value
{
  ENSURE_UI_THREAD(setTitle, value);
  ENSURE_TYPE(value, NSString);

  [[self polyline] setTitle:[TiUtils stringValue:value]];
  [self replaceValue:value forKey:@"title" notification:NO];
}

- (void)setZIndex:(id)value
{
  ENSURE_UI_THREAD(setZIndex, value);
  ENSURE_TYPE(value, NSNumber);
  
  [[self polyline] setZIndex:[TiUtils intValue:value]];
  [self replaceValue:value forKey:@"zIndex" notification:NO];
}

- (void)setUserData:(id)value
{
  ENSURE_UI_THREAD(setUserData, value);
  ENSURE_TYPE(value, NSDictionary);
  
  [[self polyline] setUserData:value];
  [self replaceValue:value forKey:@"userData" notification:NO];
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsPolylineProxy.swift
/**
 * Axway Titanium
 */

import UIKit
import TitaniumKit
import GoogleMaps
import GoogleMapsUtils

@objc(TiGooglemapsPolylineProxy)
public class TiGooglemapsPolylineProxy: TiProxy {
  private var polylineInternal: GMSPolyline?
  
  private var polylineInstance: GMSPolyline {
    if let polyline = polylineInternal {
      return polyline
    }
    
    let polyline = GMSPolyline()
    polyline.isTappable = true
    polylineInternal = polyline
    return polyline
  }
  
  private func runOnMainThread(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, false)
    }
  }
  
  private func path(from points: Any?) -> GMSMutablePath? {
    guard let points = points else { return nil }
    
    if let encoded = points as? String, let path = GMSMutablePath(fromEncodedPath: encoded) {
      return path
    }
    
    guard let collection = points as? [Any], collection.count >= 2 else {
      NSLog("[WARN] Ti.GoogleMaps: You need to specify at least 2 points to create a polyline.")
      return nil
    }
    
    let path = GMSMutablePath()
    collection.forEach { point in
      if let dict = point as? [String: Any] {
        let latitude = TiUtils.doubleValue(dict["latitude"])
        let longitude = TiUtils.doubleValue(dict["longitude"])
        path.addLatitude(latitude, longitude: longitude)
      } else if let array = point as? [Any], array.count >= 2 {
        let latitude = TiUtils.doubleValue(array[1])
        let longitude = TiUtils.doubleValue(array[0])
        path.addLatitude(latitude, longitude: longitude)
      }
    }
    
    return path
  }
  
  @objc public func polyline() -> GMSPolyline {
    return polylineInstance
  }
  
  @objc public func setPoints(_ value: Any?) {
    runOnMainThread {
      guard let path = self.path(from: value) else { return }
      self.polylineInstance.path = path
      self.replaceValue(value ?? NSNull(), forKey: "points", notification: false)
    }
  }
  
  @objc public func setDotsImage(_ value: Any?) {
    runOnMainThread {
      guard let dotsImage = value else {
        self.polylineInstance.spans = []
        self.replaceValue(NSNull(), forKey: "dotsImage", notification: false)
        return
      }
      
      guard let image = TiUtils.image(dotsImage, proxy: self) else { return }
      let strokeStyle = GMSStrokeStyle()
      strokeStyle.stampStyle = GMSTextureStyle(image: image)
      self.polylineInstance.spans = [GMSStyleSpan(style: strokeStyle)]
      self.replaceValue(dotsImage, forKey: "dotsImage", notification: false)
    }
  }
  
  @objc public func setTappable(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.isTappable = TiUtils.boolValue(value)
      self.replaceValue(value, forKey: "tappable", notification: false)
    }
  }
  
  @objc public func setStrokeColor(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.strokeColor = TiUtils.colorValue(value).color
      self.replaceValue(value, forKey: "strokeColor", notification: false)
    }
  }
  
  @objc public func setStrokeWidth(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.strokeWidth = CGFloat(TiUtils.floatValue(value, def: 1))
      self.replaceValue(value, forKey: "strokeWidth", notification: false)
    }
  }
  
  @objc public func setGeodesic(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.geodesic = TiUtils.boolValue(value, def: false)
      self.replaceValue(value, forKey: "geodesic", notification: false)
    }
  }
  
  @objc public func setTitle(_ value: Any?) {
    runOnMainThread {
      let title = TiUtils.stringValue(value)
      self.polylineInstance.title = title
      self.replaceValue(title ?? NSNull(), forKey: "title", notification: false)
    }
  }
  
  @objc public func setZIndex(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.zIndex = Int32(TiUtils.intValue(value))
      self.replaceValue(value, forKey: "zIndex", notification: false)
    }
  }
  
  @objc public func setUserData(_ value: Any?) {
    runOnMainThread {
      self.polylineInstance.userData = value
      self.replaceValue(value ?? NSNull(), forKey: "userData", notification: false)
    }
  }
  
  @objc public func setStrokeGradient(_ value: [String: Any]?) {
    runOnMainThread {
      guard let gradient = value,
            let fromColor = TiUtils.colorValue(gradient["from"])?.color,
            let toColor = TiUtils.colorValue(gradient["to"])?.color else {
        return
      }
      
      let strokeStyle = GMSStrokeStyle.gradient(from: fromColor, to: toColor)
      self.polylineInstance.spans = [GMSStyleSpan(style: strokeStyle)]
      self.replaceValue(value ?? NSNull(), forKey: "spans", notification: false)
    }
  }
}
